chatbot/chatbot_agent.py

```diff
@@ -0,0 +1,318 @@
+        REFINE_TEMPLATE = PromptTemplate(
+            input_variables=["question", "existing_anwser", "context_str"],
+            template=refine_template,
+        )
+
+        chain = load_qa_with_sources_chain(
+            # If batch_size is too high, it could cause rate limiting errors.
+            llm=OpenAI(batch_size=5, temperature=0, model_name="gpt-3.5-turbo"),
+            chain_type="refine",
+            return_intermediate_steps=True,
+            question_prompt=INITIAL_TEMPLATE,
+            refine_prompt=REFINE_TEMPLATE,
+        )
+        return chain({"input_documents": self.vectordb, "question": self.query}, return_only_outputs=True)
+
+
+    # Get the data from the merged file
+    def get_openacademysources(self, path):
+        loader = OpenAcademySourcesLoader(path)
+        data = loader.load()
+        return data
+
+
+    # Convert Markdown to Python
+    def markdown_to_python(self, markdown_text):
+        # Escape quotes and backslashes in the input
+        escaped_input = markdown_text.replace("\\", "\\\\").replace("'", "\\'")
+
+        # Generate the Python string
+        python_string = f"'{escaped_input}'"
+
+        return python_string
+
+
+    def chatbot_pipeline(self, query_pipeline, choose_GPTModel = False, updateChatHistory = False):
+        # choose which GPT model
+        if choose_GPTModel:
+            result_pipeline = openai.Completion.create(
+                engine="davinci",
+                prompt = query_pipeline,
+                temperature=0.7,
+                max_tokens=150,
+                n=1,
+                stop=None,
+            ).choice[0].text.strip() # choose the first answer whose score/probability is the highest
+        else:
+            result_pipeline = self.chatbot_qa({"question": query_pipeline, "chat_history": self.chat_history})
+        
+        if updateChatHistory:
+            self.query = query_pipeline
+            self.result = result_pipeline
+            self.chat_history = self.chat_history + [(self.query, self.reslut["answer"])]
+            return self.reslut
+        else:
+            return result_pipeline
+            
+
+
+    # Prompt the chatbot for map_reduce chain type
+    def promtp_engineering_for_map_reduce_chain_type(self):
+        self.chatbot_qchatbot_qa_retrieval_map_reduce_chain_type({"input_documents": self.vectordb, "question": self.query}, return_only_outputs=True)
+        # {'output_text': ' The president thanked Justice Breyer for his service.\nSOURCES: 30-pl'}
+
+
+    # Prompt the chatbot for non libary content
+    def promtp_engineering_for_non_library_content(self, query): # please do not modify the value of query
+        query_prompted = query + " Please provide a verbose answer."
+
+        result_prompted = self.chatbot_pipeline(query_prompted)
+        result_not_know_answer = []
+        result_non_library_query = []
+        result_official_keywords = []
+        result_cheeting = []
+        # 

```

Comment:
Please remove ALL unused code.

Reply:
Modified, but I leave the declaration, as it is still in development.